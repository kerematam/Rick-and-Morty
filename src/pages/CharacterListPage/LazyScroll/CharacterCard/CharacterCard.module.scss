$card_width: 300px;

.root {
  max-width: $card_width;
  min-width: $card_width;

  .media {
    min-height: $card_width;
    min-width: $card_width;
  }

  .content {
    padding: 16px;
    max-height: 25px;
    position: absolute;
    bottom: 0px;
    background: black;
    width: 100%;
    opacity: 0.7;
    transition: max-height 1s;
    .text {
      color: lightgrey;
    }
  }

  :hover {
    .content {
      max-height: 100%;
    }
  }
}

.detail_item {
  max-width: calc(#{$card_width} - 32px);
  display: flex;
  justify-content: space-between;
  .detail_item_key {
    color: lightgrey;
  }
  .detail_item_value {
    color: lightgrey;
    text-align: right;
    margin-left: 5px;
  }
}
