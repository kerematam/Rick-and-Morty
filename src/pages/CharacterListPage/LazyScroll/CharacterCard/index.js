import React, { useCallback } from 'react'
import { useDispatch } from 'react-redux'
import PropTypes from 'prop-types'
import { Card, CardActionArea, Typography } from '@material-ui/core'
import history from 'utils/history'
import {
  actions,
  useCharacterDetailReducerInjector,
} from 'pages/CharacterPage/redux'
import Image from 'components/Image'

import styles from './CharacterCard.module.scss'

const DetailItem = ({ label, value }) => {
  return value ? (
    <Typography className={styles.detail_item}>
      <span className={styles.detail_item_key}>{label}</span>
      <span className={styles.detail_item_value}>{value}</span>
    </Typography>
  ) : null
}

export default function CharacterCard({ character }) {
  useCharacterDetailReducerInjector()
  const dispatch = useDispatch()
  const handleClick = useCallback(() => {
    dispatch(actions.setActiveCharacter({ character }))
    history.push(`/characters/${character.id}`)
  }, [dispatch, character])

  const createdDate =
    character?.created && new Date(character.created).toDateString()

  return (
    <Card square className={styles.root}>
      <CardActionArea onClick={handleClick}>
        <Image
          className={styles.media}
          src={character?.image}
          alt={character?.name || ''}
          component="img"
        />
        <div className={styles.content}>
          <Typography className={styles.text} variant="h6" gutterBottom>
            {character?.name}
          </Typography>
          <DetailItem label="Type" value={character?.type} />
          <DetailItem label="Created" value={createdDate} />
          <DetailItem label="Origin" value={character?.origin?.name} />
          <DetailItem label="Status" value={character?.status} />
          <DetailItem label="Species" value={character?.species} />
        </div>
      </CardActionArea>
    </Card>
  )
}

CharacterCard.propTypes = {
  character: PropTypes.object,
}
