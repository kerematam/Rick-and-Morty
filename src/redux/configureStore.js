import { createStore, applyMiddleware, compose } from 'redux'
import createSagaMiddleware from 'redux-saga'
import createReducer from './rootReducers'
import { routerMiddleware } from 'connected-react-router'
import { createInjectorsEnhancer, forceReducerReload } from 'redux-injectors'

function configureStore(initialState = {}, history) {
  let composeEnhancers = compose
  const reduxSagaMonitorOptions = {}

  if (process.env.NODE_ENV !== 'production' && typeof window === 'object') {
    if (window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__)
      composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
  }

  const sagaMiddleware = createSagaMiddleware(reduxSagaMonitorOptions)
  const { run: runSaga } = sagaMiddleware
  const middlewares = [sagaMiddleware, routerMiddleware(history)]

  const enhancers = [
    applyMiddleware(...middlewares),
    createInjectorsEnhancer({
      createReducer,
      runSaga,
    }),
  ]

  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(...enhancers)
  )

  if (module.hot) {
    module.hot.accept('./rootReducers', () => {
      forceReducerReload(store)
    })
  }

  return store
}

export default configureStore
