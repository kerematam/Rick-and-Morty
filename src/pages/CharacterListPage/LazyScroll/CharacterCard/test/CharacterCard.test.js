import React, { useCallback } from 'react'
import renderer from 'react-test-renderer'
import CharacterCard from '../index'
import { Provider } from 'react-redux'
import configureStore from 'redux/configureStore'
import { createBrowserHistory, createMemoryHistory } from 'history'
import { browserHistory } from 'react-router'

const character = {
  id: 1,
  name: 'Rick Sanchez',
  type: '',
  image: '',
}

const setup = (setupProps = {}) => {
  const store = configureStore()
  const wrapper = renderer
    .create(
      <Provider store={store}>
        <CharacterCard {...setupProps} />
      </Provider>
    )
    .toJSON()

  return {
    wrapper,
  }
}

describe('CharacterCard', () => {
  test('should render correctly', () => {
    const { wrapper } = setup({ character })
    expect(wrapper).toMatchSnapshot()
  })

  test('should render without crashing with corrupted backend data', () => {
    const { wrapper } = setup({ character: { name: null } })
    expect(wrapper).toMatchSnapshot()
  })
})
