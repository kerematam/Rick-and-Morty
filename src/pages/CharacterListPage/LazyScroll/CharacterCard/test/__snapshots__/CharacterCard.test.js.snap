// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard should render correctly 1`] = `
<div
  className="MuiPaper-root MuiCard-root root MuiPaper-elevation1"
>
  <button
    className="MuiButtonBase-root MuiCardActionArea-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiSkeleton-root MuiSkeleton-rect media MuiSkeleton-wave"
      style={
        Object {
          "height": undefined,
          "width": undefined,
        }
      }
    />
    <img
      alt="Rick Sanchez"
      className="media"
      component="img"
      onError={[Function]}
      onLoad={[Function]}
      src=""
      style={
        Object {
          "display": "none",
        }
      }
    />
    <div
      className="content"
    >
      <h6
        className="MuiTypography-root text MuiTypography-h6 MuiTypography-gutterBottom"
      >
        Rick Sanchez
      </h6>
    </div>
    <span
      className="MuiCardActionArea-focusHighlight"
    />
  </button>
</div>
`;

exports[`CharacterCard should render without crashing with corrupted backend data 1`] = `
<div
  className="MuiPaper-root MuiCard-root root MuiPaper-elevation1"
>
  <button
    className="MuiButtonBase-root MuiCardActionArea-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiSkeleton-root MuiSkeleton-rect media MuiSkeleton-wave"
      style={
        Object {
          "height": undefined,
          "width": undefined,
        }
      }
    />
    <img
      alt=""
      className="media"
      component="img"
      onError={[Function]}
      onLoad={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
    />
    <div
      className="content"
    >
      <h6
        className="MuiTypography-root text MuiTypography-h6 MuiTypography-gutterBottom"
      />
    </div>
    <span
      className="MuiCardActionArea-focusHighlight"
    />
    <span
      className="MuiTouchRipple-root"
    />
  </button>
</div>
`;
